Add static vendored libraries and frameworks into fat library

manuyavuz opened the pull request:
This PR adds support for putting vendors libraries and frameworks into generated fat library while using --embedded or --library options. 

I took part of the code from PR #119. Thanks @jasl8r for the PR!

This also fixes getting correct `module_map` path when a custom `module_map` is given.

**UPDATE :** This branch also contains #195 now.

Solves #26, #49, #63, and #71.

manuyavuz committed:
Support local pods

manuyavuz committed:
Add static vendored libraries and frameworks into fat library

manuyavuz committed:
Add specs for vendored library integration

manuyavuz committed:
Update specs

manuyavuz committed:
[gem] Update gems

manuyavuz committed:
fixup! Add static vendored libraries and frameworks into fat library

manuyavuz commented:
@segiddins Not ready yet, I'll let you know.

paulb777 commented:
What does usage look like?

segiddins commented on lib/cocoapods-packager/builder.rb:
```diff
@@ -14,41 +16,52 @@ def initialize(source_dir, static_sandbox_root, dynamic_sandbox_root, public_hea
-      else
-        build_framework(platform)
+    def build(package_type)
+      if package_type == :static_library
```
this can use `case package_type` instead of all the ifs, and probably should have an `else` that raises on unexpected package type

manuyavuz replied:
I've added `:static_framework` to else case because this is what the current behavior is. If none of `--library`, `--embedded`, or `--dynamic` not given, it generates a static framework without embedding resources.

segiddins commented on lib/cocoapods-packager/builder.rb:
```diff
@@ -14,41 +16,52 @@ def initialize(source_dir, static_sandbox_root, dynamic_sandbox_root, public_hea
       platform_path.mkdir unless platform_path.exist?
-      build_library(platform, defines, platform_path + Pathname.new("lib#{@spec.name}.a"))
+
+      output = platform_path + Pathname.new("lib#{@spec.name}.a")
```
the appended component doesn't need to be wrapped in a pathname

segiddins commented on spec/specification/builder_spec.rb:
```diff
@@ -6,32 +6,34 @@ module Pod
 
         it "includes proper compiler flags for iOS" do
+          @builder = Builder.new(Platform.new(:ios), [], nil, nil, nil, nil, @spec, nil, nil, nil, nil, nil, nil)
+          puts @builder.inspect
```
`puts` in test

manuyavuz committed:
[symbols] Ignore __clang_at_available_requires_core_foundation_framework

manuyavuz committed:
Add vendored libraries of pod dependencies

This is done by passing `Installer` object to builder rather than just file accessors
of spec itself.

Builder now looks also into other pod dependencies' file assessors to find vendored
libraries.

manuyavuz committed:
Generate static frameworks by default

manuyavuz committed:
Remove unnecessary target support files copy step

manuyavuz committed:
[fixtures] Update podspec having deprecated Parse-iOS-SDK dependency

manuyavuz committed:
[fixtures] Update project contents with new CocoaPods version

manuyavuz committed:
[gem] Update gems to match `CocoaPods` repo versions

manuyavuz commented:
@paulb777 It will merge dependent static libraries by default.

But I'm not sure how it should behave when `--exclude-deps` option is given. We may choose to not merge vendor libraries of dependent pods, but what about spec's own vendored libraries? I think they should be included even if `--exclude-deps` is given, but not sure.

Suggestions on this?

manuyavuz committed:
Fix rubocop offenses

manuyavuz commented:
@segiddins, I think this is ready for review now.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

manuyavuz committed:
Add tests

manuyavuz committed:
Add test to verify local behavior

manuyavuz committed:
Check if spec is from file using `defined_in_file` method

segiddins commented:
> But I'm not sure how it should behave when --exclude-deps option is given. We may choose not to merge vendor libraries of dependent pods, but what about spec's own vendored libraries? I think they should be included even if --exclude-deps is given, but not sure.

Nah, I think the spec can continue to include them in `vendored_libraries`

manuyavuz committed:
[Builder] Remove unnecessary pathname creation

manuyavuz committed:
Do not merge vendored libraries of dependencies when --exclude-deps is given

manuyavuz committed:
Fix packaging a pod directly with name without `.podspec`

`pod package` command was failing when we run `pod package AFNetworking` because @path was nil. This should not fail, packager should generate framework using published version in one of the sources.

manuyavuz commented:
I've added ignoring vendor libraries of dependencies when `--exclude-deps` is given. Tests `includes vendor symbols both from itself and pod dependencies` and `does not include vendor symbols from pod dependencies if option --exclude-deps is specified` verify this behavior.

Gonna merge if succeeds. @segiddins, could you take one last look with approval?

manuyavuz committed:
fixup! Fix packaging a pod directly with name without `.podspec`

manuyavuz committed:
fixup! Support local pods

manuyavuz committed:
Fix failing tests

segiddins commented on spec/command/package_spec.rb:
```diff
@@ -154,17 +156,17 @@ module Pod
 
       it "includes vendor symbols both from itself and pod dependencies" do
-        Pod::Config.instance.sources_manager.stubs(:search).returns(nil)
+        # Pod::Config.instance.sources_manager.stubs(:search).returns(nil)
```
commented code

manuyavuz committed:
Remove commented code, enable test for packaging from master repo

onegaop commented:
vendors frameworks can not found headers , how can i fix it ?

lee1994522 commented:
@manuyavuz How to use it?I execute pod package MyProject.podspec --force --embedded but still can't find vendored_framework merged into the package.Here is part of my podspec:
`
    s.subspec 'LockstepApi' do |ss|

    ss.ios.source_files = 'JSBridgeSDK/LockstepApi/**/*.{h,m,c,mm,inc}'

    ss.ios.public_header_files = 'JSBridgeSDK/LockstepApi/**/*.h'

    ss.ios.vendored_frameworks = 'JSBridgeSDK/Frameworks/*.framework'

    end
`
But JSBridgeSDK/Frameworks/*.framework still not merged

lee1994522 commented:
So how can i use this feature?Just by using '--embedded' option?

gaojiefeng commented:
> 

same question?have you get any idea yet?

jaykon commented:
> same question?have you get any idea yet?

I guess you using 1.5.0 witch is a old version，and this feature is update at master without a tag，you should clone from master and install it.

jaykon commented:
@gaojiefeng @lee1994522 

I guess you using 1.5.0 wich is an old version，and this feature is update at master without a tag，you should clone from master and install it. (guide : https://www.jianshu.com/p/a253017eecce )

guanzhendong commented:
@jaykon 我按照你说的方式，安装了master最新分支，但是还是不能在依赖framework情况下，使用mangling，还是打包失败，看来还是没有解决这个问题
